src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1318 @@
+   MONGOC_THREAD_SHUTTING_DOWN,
+   MONGOC_THREAD_JOINABLE
+} thread_state_t;
+
+/* Use a signed and wide return type for timeouts as long as you can. Cast only
+ * when you know what you're doing with it. */
+static int64_t
+_now_us (void)
+{
+   return bson_get_monotonic_time ();
+}
+
+static int64_t
+_now_ms (void)
+{
+   return _now_us () / 1000;
+}
+
+struct _mongoc_server_monitor_t {
+   mongoc_topology_t *topology;
+   bson_thread_t thread;
+
+   /* State accessed from multiple threads. */
+   struct {
+      bson_mutex_t mutex;
+      mongoc_cond_t cond;
+      thread_state_t state;
+      bool scan_requested;
+      bool cancel_requested;
+      mongoc_interrupt_t *interrupt;
+   } shared;
+
+   /* Default time to sleep between ismaster checks (reduced when a scan is
+    * requested) */
+   uint64_t heartbeat_frequency_ms;
+   /* The minimum time to sleep between ismaster checks. */
+   uint64_t min_heartbeat_frequency_ms;
+   int64_t connect_timeout_ms;
+   bool use_tls;
+#ifdef MONGOC_ENABLE_SSL
+   mongoc_ssl_opt_t *ssl_opts;
+#endif
+   mongoc_uri_t *uri;
+   /* A custom initiator may be set if a user provides overrides to create a
+    * stream. */
+   mongoc_stream_initiator_t initiator;
+   void *initiator_context;
+   int64_t request_id;
+   mongoc_apm_callbacks_t apm_callbacks;
+   void *apm_context;
+
+   mongoc_stream_t *stream;
+   bool more_to_come;
+   mongoc_server_description_t *description;
+   uint32_t server_id;
+   bool is_rtt;
+};
+
+#ifdef MONGOC_TRACE
+static void
+_server_monitor_log (mongoc_server_monitor_t *server_monitor,
+                     const char *format,
+                     ...) BSON_GNUC_PRINTF (2, 3);
+
+static void
+_server_monitor_log (mongoc_server_monitor_t *server_monitor,
+                     const char *format,
+                     ...)
+{
+   va_list ap;
+   char *msg;
+
+   va_start (ap, format);
+   msg = bson_strdupv_printf (format, ap);
+   va_end (ap);
+
+   TRACE ("[%s%s] %s",
+                 server_monitor->description->host.host_and_port,
+                 server_monitor->is_rtt ? "-RTT" : "",
+                 msg);
+   bson_free (msg);
+}
+
+#define MONITOR_LOG(sm, ...) _server_monitor_log (sm, __VA_ARGS__)
+#else
+#define MONITOR_LOG(sm, ...) (void) 0
+#endif
+
+/* Called only from server monitor thread.
+ * Caller must hold no locks (user's callback may lock topology mutex).
+ * Locks APM mutex.
+ */
+static void
+_server_monitor_heartbeat_started (mongoc_server_monitor_t *server_monitor,
+                                   bool awaited)
+{
+   mongoc_apm_server_heartbeat_started_t event;
+
+   if (!server_monitor->apm_callbacks.server_heartbeat_started) {
+      return;
+   }
+
+   event.host = &server_monitor->description->host;
+   event.context = server_monitor->apm_context;
+   MONITOR_LOG (server_monitor, "%s heartbeat started", awaited ? "awaitable" : "regular");
+   /* TODO: once awaitable support on heartbeat events is merged. */
+   /* event.awaited = awaited; */
```

Comment:
After https://github.com/mongodb/mongo-c-driver/pull/621 is merged, this will be rebased, uncommented, and tested.